engine/src/main/scala/cromwell/engine/workflow/WorkflowStoreActor.scala

```diff
@@ -0,0 +1,145 @@
+  * changes, revisit.
+  */
+trait WorkflowStore {
+  import WorkflowStore._
+
+  /*
+    The LinkedHashMap preserves insertion order which means that any requests for runnable workflows will
+    be satisfied by taking the oldest available.
+   */
+  private var workflowStore = new mutable.LinkedHashMap[WorkflowId, SubmittedWorkflow]
+
+  /**
+    * Adds a WorkflowSourceFiles to the store and returns a WorkflowId for tracking purposes
+    */
+  def add(source: WorkflowSourceFiles): WorkflowId = {
+    val workflowId = WorkflowId.randomId()
+    workflowStore += (workflowId -> SubmittedWorkflow(source, Submitted))
+    workflowId
+  }
+
+  /**
+    * Provides a bulk add functionality. Adds all of the provided WorkflowSourceFiles at once, returning a List
+    * of UUIDs matching the order of the supplied source files
+    */
+  def add(sources: Seq[WorkflowSourceFiles]): immutable.List[WorkflowId] = {
+    val submittedWorkflows = sources.toList map { WorkflowId(UUID.randomUUID()) -> SubmittedWorkflow(_, Submitted) }
+    workflowStore ++= submittedWorkflows
+    submittedWorkflows map { _._1 }
+  }
+
+  /**
+    * Retrieves up to n workflows which have not already been pulled into the engine and sets their pickedUp
+    * flag to true
+    */
+  def fetchRunnableWorkflows(n: Int): immutable.List[WorkflowToStart] = {
+    val startableWorkflows = workflowStore filter { case (k, v) => v.state.isStartable } take n
+    workflowStore ++= startableWorkflows map { case (k, v) => k -> SubmittedWorkflow(v.sources, Running) }
+    // Because we explicitly filtered out for startable states this is safe
+    startableWorkflows map { case (k, v) => v.toWorkflowToStart(k) } toList
+  }
+
+  // FIXME: Do we care if it doesn't exist? If so to what extent
+  def remove(id: WorkflowId): Unit = workflowStore -= id
+
+  protected def dump: immutable.Map[WorkflowId, SubmittedWorkflow] = workflowStore.toMap
+}
+
+class WorkflowStoreActor extends WorkflowStore with Actor with ServiceRegistryClient {
+  /*
+    WARNING: WorkflowStore is NOT thread safe. Unless that statement is no longer true do NOT use threads
+    outside of the the single threaded Actor event loop
+   */
+
+  private val logger = Logging(context.system, this)
+
+  override def receive = {
+    case SubmitWorkflow(source) =>
+      val id = add(source)
+      sendIdToMetadataService(id)
+      sender ! WorkflowSubmitted(id)
+    case SubmitWorkflows(sources) =>
+      val ids = add(sources)
+      ids foreach sendIdToMetadataService
+      sender ! WorkflowsSubmitted(ids)
+    case FetchRunnableWorkflows(n) => sender ! NewWorkflows(fetchRunnableWorkflows(n))
+    case RemoveWorkflow(id) => remove(id)
+  }
+
+  /**
+    * Takes the workflow id and sends it over to the metadata service w/ default empty values for inputs/outputs
+    */
+  private def sendIdToMetadataService(id: WorkflowId): Unit = {
```

Comment:
Agreed, although the one I changed it to you might not like any better, but you'll need to wait for my next push to see it